### ISSUE TITLE
Working in multiple domains in the same forest

### ISSUE THREAD

**Issue by iyovkov** [lithnet/laps-web#10]
I am having difficulties making LAPSweb working in a multi domain forest environment. 

I have DomainA.contoso.com and DomainB.contoso.com. LAPS and LAPSweb are installed in DomainA. I got it working for this domain. I have an OU containing all client computers in DomainB. Already gave read permissions for the service account on the OU. 

First I tried to set the OU from DomainB in the web.conf and add the designated security group (which is from DomainA) as a reader, but this didn't work out. Then I directyl set the "Domain Computers" group there and left the same group. Still no luck.

   <targets> 
	<target name="Domain Computers" 
            type="Group" 
            expire-after="02:00:00">
       <readers>
         <reader principal="DomainA\SecurityGroup" />
       </readers>
    </target>
  </targets>

LAPS is working fine and I can see the passwords from the GUI or from Powershell in both domains, so no issues on this side.

Is this possible or am I missing something?

**Comment by bezmena** [comment]
Hey, i am having the same problem. Have you found out any solution yet?

**Comment by ryannewington** [comment]
Hi guys,

My apologies but I'm on leave at the moment and have very little internet access. There is likely an assumption in the code of a single domain. I suspect we are binding to the DC in a site the LAPS server is joined to, which would mean queries against other domains would 'fail' with a referral.

I'll need to tackle this one when I get back, but that's not going to be until late June unfortunately.

Ryan

**Comment by bezmena** [comment]
You are right, I tried to dig in and changed `PrincipalContext` in `GetComputerPrincipal` function like this:
```c-sharp
PrincipalContext ctx = new PrincipalContext(ContextType.Domain, "server:3268", "rootdomain");
```
where `server` is any address of global catalog running at port 3268
That will find any computer in given forest.

**Comment by fsecchia** [comment]
Hi there, exactly the same problem here. Any chance to have a new build supporting global catalog and domain forest? 

Thanks
Fabio

**Comment by iyovkov** [comment]
Hello guys. I tried what bezmena suggested and it worked out. Thanks a lot! I now discovered one more issue. I can't use accounts form different child domains. Currently I set it up to work with Windows authentication. Whenever I try logging with an account form a different domain I get this error:

Your request could not be processed because your SSO identity could not be found in the directory
System.DirectoryServices.AccountManagement.NoMatchingPrincipalException: The request could not be processed because the authenticated user was not found in the directory. Claim information:
XXX\XXXXXX
   at Lithnet.Laps.Web.Controllers.LapController.GetCurrentUser() in d:\github\lithnet\laps-web\src\Lithnet.Laps.Web\Lithnet.Laps.Web\Controllers\LapController.cs:line 225
   at Lithnet.Laps.Web.Controllers.LapController.Get(LapRequestModel model) in d:\github\lithnet\laps-web\src\Lithnet.Laps.Web\Lithnet.Laps.Web\Controllers\LapController.cs:line 43

I tried going trough the source code, but unfortunately, C# is not my thing :(

Thanks!

**Comment by red-erik** [comment]
Hello,
@bezmena: would you be so kind to give more details ? Im not a programmer and I don't know where to find GetComputerPrincipal function. The same I can't recompile source code.
Thank you in advance.

Regards.

Red.

**Comment by ryannewington** [comment]
Hi all,

I'm currently waiting on a dev environment to be spun up with multiple domains, forests, and trusts, so I can test and make sure all scenarios are working. I appreciate everyone's patience on this issue, and rest assured, it is the first feature out the door with the next version.

Ryan

**Comment by red-erik** [comment]
Thank you for the update.
Anyway, if more simple, it would be enough to create anew Authentication method ad STANDARD LDAP, passing LDAP server and port for auth and object search. That way should be possible to use a global catalog.
Regards.

Red.

**Comment by ryannewington** [comment]
@red-erik I'm happy to put something together, but I can't test it at the moment. Do you have the capability to validate the change and ensure it works as you expected?

**Comment by red-erik** [comment]
Sure. I'm working in a corporate environment with a single Active directory with 1 root domain and 4 child domains...

**Comment by dlsmlss** [comment]
@ryannewington also can join to "lab rats club"

**Comment by ryannewington** [comment]
Thanks guys.

Can you also let me know where objects are located in your setup. Which domain are the users in, which domain are the computers in, where your security groups are located, and where the LAPS server is located. A simple little diagram would be awesome. 

Ryan

**Comment by red-erik** [comment]
LAPS server on root Domain (Domain A) while users and computer in 4 sub domains (Domain B, Domain C, Domain D and Domain E). Security groups are present in ALL domains.
It's a simple AD tree in witch from Root you find OU,Users, Groups, Computers and **subdomains** and in Subdomain you find OU, Users, Groups, Computers....(bi-direction implicit trust relationship are in place, by default).

Red.
![ad](https://user-images.githubusercontent.com/10245160/65970751-617f2e00-e467-11e9-8a80-183504684cba.jpg)

**Comment by dlsmlss** [comment]
Almost same scheme like @red-erik has
![diag](https://user-images.githubusercontent.com/41381305/66031186-2c281e00-e503-11e9-808a-8b4867ba512a.png)
laps-web computer joined to root domain.
IT stuff from each subdomain request passswords for own domain.

**Comment by red-erik** [comment]
Hello,
any news an some new bit to test ? Please remember that I can't compile source code so to able to test it I need a pre-compiled version.
Regards.

Red.

**Comment by ryannewington** [comment]
Hi guys, thanks for getting those to me. Very helpful. I should have something for you to test this weekend.

Ryan

**Comment by ryannewington** [comment]
Ok here is the test build.

[Publish.zip](https://github.com/lithnet/laps-web/files/3720031/Publish.zip)

Backup your web.config files and deploy this version over the top of your current web app.

I've tested this against a single-forest GC, and it seems to be working ok. I don't have my full multi-domain environment set up yet, so I'll admit, some of this has been reading MDSN docs and guess work.

One thing that was tricky for me to test, and I don't really know the exact behaviour we will have is group membership. If you could test on group-based reader prinpicals both in and out of the domain where LAPS is installed. I'd be intereested to hear what scenarions work and do not work, and which are expected based on your previous cross-domain group experience.

There are no config settings to turn on GC support, as all lookups are now GC based.

Sorry this has taken so long to get to!

**Comment by ryannewington** [comment]
A couple of notes on this release

- You'll need .NET 4.8 installed
- Overwrite the following blocks in the web.config file with those from your old file that you backed up 
  * `<lithnet-laps>`
  * `<appSettings>`
  * `<nlog>`
  * `<system.net>` if you are using the email functionality

**Comment by dlsmlss** [comment]
@ryannewington thanks Ryan! Will try to test it on Mon-Wed.
Btw, @red-erik asked you about LDAP/LDAPS 3268/3269. Is it implemented in this release?

**Comment by ryannewington** [comment]
The work around suggested providing the ability to set a global catalogue ldap server and Port is no longer needed.

We use GC for all searches now, and then grab objects from the domains they reside in.

**Comment by red-erik** [comment]
Hello,
thank you for your time and fo a new relese.
Those are the result from my test:
User in root domain and server in root domain --> OK
User in child domain and pc in child domain --> see LAPS2.jpg
User in root domain and pc in child domain  --> seel LAPS1.jpg

Basically, it seems the the AD group membership (used to verify if the user can access LAPS password) and computers "AD container" are still not managed correctly.

If can enable a verbose log, I'll send you the detailed information.

Regards,
Red.

**Comment by ryannewington** [comment]
@red-erik can you send me the log file from the app_data/logs folder - ryan@lithnet.io

**Comment by dlsmlss** [comment]
@ryannewington Hi Ryan!
I have finished my tests and can confirm that in my case everything working as expected. Tested scenarios:
- OU in child - requester in root domain
- OU in root - requester in child
- Group in child - requester in root
- Group in child - requester in another or same child

PS: also i made "clean" install on W2019 and discovered that you have to select Static content in IIS or some disign may be "corrupted" and you cant see logo for example. So:
`Install-WindowsFeature -Name Web-Static-Content`

**Comment by dlsmlss** [comment]
@red-erik btw... did you add permissions with command suggested by Ryan?
Have same errors but after that I used 
```
Set-AdmPwdReadPasswordPermission -Identity "OU=SERVERs,DC=us,DC=xxxx,DC=local" -AllowedPrincipals "xxxx\gmsa_laps$"
Set-AdmPwdResetPasswordPermission -Identity "OU=SERVERs,DC=us,DC=xxxx,DC=local" -AllowedPrincipals "xxxx\gmsa_laps$"
```
and it fixed my problem

**Comment by ryannewington** [comment]
@dlsmlss great news. What type of authentication are you using? Windows auth works a little differently from OIDC and WS-Fed, so just want to know which scenario is confirmed working for you.

@red-erik what auth mode are you using?

**Comment by dlsmlss** [comment]
@ryannewington Ryan, my test done with ADFS auth.

**Comment by red-erik** [comment]
Hello,
I've checked the log and it simply states that I'm not authorized to view computer password while I've explicitly authorized an AD User Group to read it with Set-AdmPwdReadPasswordPermission.
This morning I'll recheck everything but if I use Fat Client everything is working as expected so it's not a permission problem.

@ryannewington : Does the app search for computer objects in the specified container and sub containers or only on the root container?

Regards.

Red.

**Comment by ryannewington** [comment]
@red-erik

Is a bit hard to guess what's going on. I need to see the logs myself. Are you able to share them, as well as your Web.config file? You can email them to ryan@lithnet.io

Ryan

**Comment by dlsmlss** [comment]
I dont know how it's related to @red-erik  problem, by have something similar...
I have those lines in config.
```
      <target name="MSS-EXMB01" type="Computer" expire-after="01:00:00">
        <readers>
		  <reader principal="RU\username" />
        </readers>
      </target>
      <target name="OU=SERVERs,DC=xxxx,DC=local" type="Container" expire-after="01:00:00">
        <readers>
		  <reader principal="XXXX\Domain Admins" />
        </readers>
      </target>
```
Computer MSS-EXMB01 belongs to "OU=SERVERs,DC=xxxx,DC=local" container.
When I try to get password by requester username i got "You are not authorized to access the password for this computer"
In logs:
```
2019-10-14 18:55:37.3314| INFO|username has requested the password for mss-exmb01
2019-10-14 18:55:37.4304|TRACE|Matched MSS-EXMB01$ to target OU OU=SERVERs,DC=xxxx,DC=local
2019-10-14 18:55:37.5374|ERROR|AUDIT: FAILURE
```

**Comment by ryannewington** [comment]
@dlsmlss can you send me the rest of the audit log? Email is fine if you don't want to paste on here

**Comment by dlsmlss** [comment]
@ryannewington 📧 sent.

**Comment by ryannewington** [comment]
Updated build to fix issue reported by @dlsmlss. There was an error in the code impacting computer-based target types. 

[Publish.zip](https://github.com/lithnet/laps-web/files/3728392/Publish.zip)

**Comment by dlsmlss** [comment]
@ryannewington I can confirm that computer related problem fixed, but some strange things with "cross-targets" still present.

**Comment by ryannewington** [comment]
The code that dealt with target prioritization was not working correctly. Computer targets should take preference over Container targets, but that didn't happen in this case. 

Updated build to fix this issue: 
[Publish.zip](https://github.com/lithnet/laps-web/files/3728669/Publish.zip)

**Comment by red-erik** [comment]
Hello,
for me the same problem...FAT UI working, laps web not. The account used for Application Pool is an enterprise admin and it is part of all authorized groups. See jpeg

Regards,

Red.
![laps_3](https://user-images.githubusercontent.com/10245160/66824378-728c6c80-ef48-11e9-93c0-3a0de9c8c70b.JPG)

**Comment by red-erik** [comment]
Ok solved it..it's time to search for a new job...I simply wrote the wrong group in web.config.
I'll do all possible tests today afternoon and let you know the results.
Forgive my foolish error.

Regards,

Red.

**Comment by dlsmlss** [comment]
> Computer targets should take preference over Container targets, but that didn't happen in this case.

Am I right that priority is:
- Standalone computer
- Group
- Organizational unit 

Anyway, I can confirm that all working as expected now. 🍻

**Comment by red-erik** [comment]
Another question is:
how search is made in AD ? We have a lot of OU nested containing client PC. I have authorization to readl LAPS password on ALL of them but in web.config, do I need to specify them all or only the base DN ?

Regards,

Red.

**Comment by dlsmlss** [comment]
@red-erik root DN of OU will be enought.

**Comment by red-erik** [comment]
Thank you @dlsmlss . I'm checking it right now

**Comment by red-erik** [comment]
Hello,
I have the setup working but with an exception:
User in Subdomain A can't read password for PC in SubdomainB. Obviously the Fat Client is working.
All other tests are fine and everything else is working as expected.
Regards,

Red.

**Comment by ryannewington** [comment]
@red-erik Can you outline how the user is a member of that group and the group types . Eg
Domain1\user is a member of domain2\somegroup (global group) which is a member of domain3\lapsgroup (universal group)

Can you also add your laps service account to the 'Windows authorisation access' group and see if that resolves the issue.

**Comment by red-erik** [comment]
Hello,
this is the architecture:
Red member of domainA\LAPS_Readers (Domain Local Group): able to read password from domainA computers.
Red (from domainA) member of domainB\LAPS_Readers (Domain Local Group): unable to read password for domainB computers with error |TRACE|Current user Red is not a member of the group reader principal domainB\LAPS_Readers

The account used to to run LAPS Application Pool is an Enterprise Admin of the ROOT domain (so member of Domain Admins in ALL subdomains).

Red.

**Comment by ryannewington** [comment]
@red-erik your group membership is calculated in "domain a" when your account lives. Domain local groups in other domains are not part of this membership list. Your "domainb\lapsreader" group needs to be a global group, so it appears in the group membership logon token for your account in "domain a".

As a side note: some unsolicited advice. Run laps as a standard user account (or gsma) not a da/ea. It doesn't need that level of access. Grant it rights to read laps password and set expiry date only.

**Comment by red-erik** [comment]
Hello,
I'm trying running LAPS with Enterprise Admin only for testing purpose but I'll change it to a standard account with required permission.
We need to use local domain groups because we need to populate them with account from a trusted forest delegated to AD management and nesting requirements.
The FAT Client works as expected so I assume that we could change how LAPS Web application check for group membership. In a direct query to "who is member of domainB\LAPS_readers group", my name is always present.

Red.

**Comment by ryannewington** [comment]
If you run the laps client from a machine in domain A as the user from domain A, are you able to read the laps password for a machine in domain B?

Or are you running the laps client as the user from domain A, but on a machine in domain B?

**Comment by red-erik** [comment]
I run Fat client from domainA, on mu PC indomainA with my account in domainA and I'm able to read pc password for a machine in domainB

**Comment by ryannewington** [comment]
If you run "whoami /groups" on the pc in domain A, do you see the "domainB\LAPS_readers" group?

**Comment by red-erik** [comment]
No, I can see membership in other domains' groups but I DON'T see the doaminB\LAPS_Readers

**Comment by ryannewington** [comment]
Ok that's what I would have expected. That list is the same list we use to check permissions. 

Could there be another group that is actually giving you permission to read the laps password in the other domain? One that uses a universal group instead?

**Comment by red-erik** [comment]
not, for sure. My personal account is a standard Ad user without any particular privilege, I only receive permissions through Group membership

**Comment by red-erik** [comment]
Hello,
as written before, I can't change that group because it has members from a TRUSTED Active Directory Forest used for AD management so ADUC won't let you change it without removing foreign groups. Our security structure REQUIRES that we manage production AD from a foreign trusted dedicated forest.

**Comment by red-erik** [comment]
anyway, Fat client work fine and fast, in a few seconds I have the password but I don't know how it check for permissions..probably it simply try to read the password of the computer using my account

**Comment by red-erik** [comment]
This could and I say "COULD" be an idea. LAPS access is already managed and protected by permissions in AD, so, why we need to double check if the user is authorized or not ? wouldn't be simpler to pass the browser credential directly to the read request and grab the results ?
I'm not a programmer and I don't know if it easy to do or the most complicated thing to implement....

**Comment by red-erik** [comment]
OK..got it working. In readers section of web.config I need to put the LAPS_Readers group from ALL involved domains, eg:
readers>
          <reader principal="A_domain\Admin_LAPS_PC" />
	  <reader principal="B_domain\Admin_LAPS_PC" />
          <reader principal="C_domain\Admin_LAPS_PC" />
</readers>

**Comment by red-erik** [comment]
but from Foreign Ad the problem is still here. The security token solution is suitable for a single Active Directory and not Multiple Active Directory's forests connected with TRUST
@ryannewington  Thank you very much for your time and help

Regards,

Red.

**Comment by ryannewington** [comment]
Let me have a look at the domain local scenario further.

Regarding cross-forest trusts, that's going to have to wait until I have my new dev environment set up. I can't do much without that.

**Comment by red-erik** [comment]
sure, and thank you for your valuable work.

Red.

**Comment by ryannewington** [comment]
Ok,

So the reason that the fat client is working turns out to be that while your logon token is generated in your home domain, and does NOT have that domain local groups from the other domain in it, when its presented to the target server, a new impersonation token is created by the server authorizing you to read the LDAP directory. Because that server is in the other domain that includes the local groups, your new impersonation token contains your domain local groups from that domain, hence the access to the LAPS password works.

As we are not creating a logon token and passing it to the other domain on your behalf, we don't get your domain local groups automatically. There is an API that can be called to generated your token groups against another domain, so I've attempted to use that here. However, that API only exists on server 2012 and above, and my development environment is 2008r2. So I haven't been able to test the remote authorization calculation, but I can and have tested it locally.

Here's a build with what I believe should work, but if it doesn't, we may have to wait for my new dev environment. This stuff is quite complicated to get right, and I'm flying blind in my current environment.

[Publish.zip](https://github.com/lithnet/laps-web/files/3742711/Publish.zip)

**Comment by red-erik** [comment]
Hello,
thank you for this. My LAPS Web application is running on Windows 2016 server. I updated with new zip but if I remove the groups from other child domain from web.config (I did not override it, should I ?) it does not work for the usual reason: Matched reader principal empty.
Have I to configure something else?

Red.

**Comment by ryannewington** [comment]
@red-erik Do you mind sharing the relevant log section for the attempt? The end result only tells me what happened (it didn't work), but doesn't tell me why. The log shows me the steps the code went through, and from that I can infer what part of the resolution process failed.

**Comment by red-erik** [comment]
Ok..I see the new BOX for the request reason but we have a couple of problem:
1) unable to connect to remote server to request token
2) something wrong in the email sending sequence (I receive the email but the system is trying to send email to foreign address too, not allowed by our SMTP server)
3) we have a lot of firewalls between servers so, witch port do you need to use to reach another child domain through API ?
4) The domain controllers ARE Windows 2008 R2 but the server running LAPS is Windows 2016

2019-10-18 12:36:04.1382| INFO|Redi has requested the password for RM1HPD600027 
2019-10-18 12:36:04.3726|TRACE|Matched RM1HPD600027$ to target OU OU=RM-Computers,DC=rm,DC=aaaaa,DC=bb 
2019-10-18 12:36:04.4195|TRACE|Reader principal rmaleniaspazio\Admin_LAPS_PC found in directory as user CN=admin_LAPS_PC,OU=AD Administration,OU=RM-Secad,DC=rm,DC=aaaaa,DC=bb 
2019-10-18 12:36:04.4820| WARN|Unable to connect to the remote server rm.aaaaa.bb to generate the authorization token for principal S-1-5-21-2531055191-855916945-2278501704-9390. The local server will be used instead, however the token generated may not contain authorization groups from other domains System.ComponentModel.Win32Exception (0x80004005): **_There are no more endpoints available from the endpoint mapper_**
   at Lithnet.Laps.Web.ActiveDirectory.Interop.NativeMethods.<GetTokenGroups>d__26.MoveNext() in D:\dev\git\lithnet\laps-web\src\Lithnet.Laps.Web\Lithnet.Laps.Web\ActiveDirectory\Interop\NativeMethods.cs:line 332
2019-10-18 12:36:04.5132|TRACE|Reader principal aaaaaa\Admin_LAPS_PC does not match current user Red 
2019-10-18 12:36:04.5132|TRACE|User Red did not match any reader principal rules for target OU=RM-Computers,DC=rm,DC=aaaaa,DC=bb 
2019-10-18 12:36:04.5601|ERROR|AUDIT: FAILURE          
Username:			Red
Name:				Red
UPN:				Red@aaaaa.bb
SID:				S-1-5-21-2531055191-855916945-2278501704-9390
....."omissis"
Reader principal:			
Detail:				The user Red is not authorized to access the computer RM1HPD600027.
Request reason:	    	test new api 
2019-10-18 12:36:10.0368|ERROR|An error occurred sending the failure audit email System.Net.Mail.SmtpFailedRecipientException: Mailbox unavailable. The server response was: 5.7.54 SMTP; **Unable to relay recipient in non-accepted domain**
   at System.Net.Mail.SmtpClient.Send(MailMessage message)
   at Lithnet.Laps.Web.Mail.SmtpMailer.SendEmail(IEnumerable`1 recipients, String subject, String body) in D:\dev\git\lithnet\laps-web\src\Lithnet.Laps.Web\Lithnet.Laps.Web\Mail\SmtpMailer.cs:line 47
   at Lithnet.Laps.Web.Audit.Reporting.PerformAuditFailureActions(LapRequestModel model, String userMessage, Int32 eventID, String logMessage, Exception ex, ITarget target, AuthorizationResponse authorizationResponse, IUser user, IComputer computer) in D:\dev\git\lithnet\laps-web\src\Lithnet.Laps.Web\Lithnet.Laps.Web\Audit\Reporting.cs:line 87

**Comment by ryannewington** [comment]
@red-erik apologies I must have left my test Web.config in there. You can overwrite that fully with your own. There's no changes needed from the last version you had.

Unfortunately though, the dc would need to be 2012 or higher for this new method to work. The error we got is because the rpc service to do the authz check doesnt exist in that version of windows. 

Do you have any 2012+ domains with local groups that we could use as a test at all?

**Comment by red-erik** [comment]
Unfortunately on production Domains ALL DC are Windows 2008 R2. We will upgrade to Windows 2016 but not immediately.
Really thank you for all you support and great work.
I'll roll back to previous version with Groups from all domain configured in the web.config because that way it works.
Can you tell me how to preserve the REASON BOX (in email too) with previous version ?
Thank you very much.

Red.

**Comment by ryannewington** [comment]
The reason stuff is still being developed, you can keep an eye on this issue and as soon as I finalise the config settings I'll post them there https://github.com/lithnet/laps-web/issues/11

**Comment by red-erik** [comment]
Hello,
any news on this topic ?

Regards,
Red.

**Comment by ryannewington** [comment]
Hi all,

A quick update. I've been doing a heap of work on v2 of this product, with a bunch of new features, including getting this multi domain scenario sorted out.

I've got my test environment, and have fixed a few problems in the parent/child domain scenario and also with two-way trusted forests. That's all working fine now.

@red-erik I can see the problem you are having with the one-way trusted forest. 

It is a difficult problem to solve due to the way the windows authorizarion APIs are behaving. When we have laps in the trusted (admin) forest, we call into the trusting forest (prod) to ask it to build the authorization information for the user (from admin). The trusting (prod) DC is then trying to make a call to the trusted DC (admin) which is not allowed and is (correctly) getting denied. This causes the whole process to fail.

I'll try opening a case with Microsoft to see if they can shed some light on this strange behaviour. Otherwise I'll need to find another solution for the one-way forest trust scenario. 

I am committed to supporting this scenario as I will need it for my own environment soon.

**Comment by ryannewington** [comment]
@red-erik, can I confirm the scenario you are using here?

1. RED forest is trusted by GREEN forest, but not the other way around
2. Domain Local group in GREEN forest grants access to LAPS Passwords in green forest

Where are your user accounts located? In the GREEN or RED forest?

The scenarios I have at the moment (with LAPS running from within the RED forest):
**WORKING** - User in RED forest, who is a member of a domain local group in RED forest can access LAPS password for computer in GREEN forest
**WORKING** - User in GREEN forest, who is a member of a domain local group in GREEN forest can access LAPS password for computer in GREEN forest
**NOT WORKING** - User in RED forest, who is a member of domain local group in GREEN forest cannot access LAPS password for computer in GREEN forest

So essentially, as long as the group and user are in the same domain, we have a solution. Will this work for your use case?

**Comment by red-erik** [comment]
Hello,
I'm in the NOT WORKING scenario. I mean, I can manage group membership as I want but, basically, users in RED forest are part of Groups in RED forest. Those groups (in RED forest) are members of groups in the GREEN forest (this is a security constrain to have a Management FOREST and a Managed FOREST) that are authorized to read LAPS passwords.
Glad to help you in testing V2 version, even if in "beta" phase.

Red.

**Comment by ryannewington** [comment]
Hi red,

If the groups exist in the RED forest, and are members of groups in GREEN forest, can we not just target the RED groups instead of green?

Ryan

**Comment by red-erik** [comment]
Hello,
this is where we started from. I can't use different forest users to correctly authenticate in LAPS Web interface. It works using FAT Client but no in the Web interface. The same point is that I need to authorize all groups from all subdomains to have it working correctly (Global Catalog, etc.etc.)

Red.

**Comment by ryannewington** [comment]
Hey Red,

Sorry mate I don't mean to frustrate you. It's very complicated to build this stuff, and I'm trying to make sure I have solutions that work for you. At the end of the day, I can't make it work the same way the LAPS client does. In the LAPS client scenario, the evaluation happens on a directory ACL in the GREEN domain. I can't do that as it stands today.

The difference is, I don't have the user's windows logon token.  When you access the LAPS password using the fat client, A logon is being performed in the GREEN domain, which evaluates the token groups differently as to how they are seen in the RED forest. I don't have the user's credentials, so I can't do that as simply as in the fat client scenario.

So I can do (1) GREEN user in GREEN group accessing GREEN computer. Or (2) RED user in RED group accessing GREEN computer, but I can't do (3) RED user in GREEN group accessing GREEN computer.

This only applies to cases where there is a one-way trust (GREEN trusts RED). In all other scenarios (child domains, two way trusts, etc) there are no problems and it all works.

So if I understand correctly, you are saying that you cannot configure your environment in such a way to make (1) or (2) work? If that's the case, can you indulge me a little further and help me understand the requirements, so I can set this up in my lab, and I'll probably have to build something unique to accommodate this scenario? Whatever it will be, it's a lot of work, and I need to understand the requirements properly.

Ryan

**Comment by red-erik** [comment]
Hello Ryan,
I understand your point and I don't want you to rebuild everything to accomplish our needs. I was only suggesting that "it would be fine" to have this configuration working because it's quite diffused (based on MS best practice) to have a Management Directory with a One way trust to the Production Directory.
Anyway, the solution 2 is clearly suitable, we only need to re-run the scripts to assign permissions to Groups in RED domain too and not only to Groups in GREEN domain.
The fact the child domain with two way trust is now fully functional is a great improvement (simplifying a lot the permission scheme).
Thank you again for all your great work and to share it with the community.

Regards,

Red.

**Comment by ryannewington** [comment]
@red-erik thanks for getting back to me. I'm definitely committed to making sure the tool works in all environments. We are also moving to ESAE/red forest model in our environment, so that scenario is key to support. But yeah, we're a fundamentally different application to the native client, and we have to work within the bounds of what Windows allows us to do. The important thing is that we have a way to achieve the access we need, even if that means a different configuration than what is needed to support the native client.

I'm a few weeks away from have the beta ready to test, and i'd really appreciate your feedback when we get to that point. It's all coming along really well, lots of new features, a new configuration user interface, so we don't have to deal with xml files by hand anymore.

**Comment by PrzemyslawKlys** [comment]
Quick question - Does current release (public) supports a single forest with multiple domains? I saw that there were some 'test versions' uploaded to this thread just not sure if it has been already part of the current build and you're just fighting now with cross-forest scenarios?

**Comment by ryannewington** [comment]
@PrzemyslawKlys To be honest, I've lost track of what build fixed what. You are good to use the build linked in this comment. This is used in production in my own environment.

https://github.com/lithnet/laps-web/issues/10#issuecomment-543568975

That build will work for the scenario you described. The complicated scenario is in one way trusts with external forests.

V2 will be out for testing soon which include support for all these scenarios and more. It's a major release and ive been investing all efforts for cross forest scenarios into this release.

**Comment by PrzemyslawKlys** [comment]
Cool, I'll wait for that one then. I can help you test it on cross-forest one if needed as I have like 5 forests connected to one, tho with Twoway trusts. 

Btw. I was thinking to write PS module to allow easy editing (adding/removing) the XML file - but as I read your message it means there will be different interface for adding everything? Should I wait or when my Client will want this feature write it?

**Comment by ryannewington** [comment]
@PrzemyslawKlys Yes it's a full user interface for configuring all settings and permissions. It's a different json-based config file as well. The new app runs on dot net core and does away with IIS and instead runs on http.sys. Quite a lot has changed. It also provides a powershell based rule to allow you to check an external system for permission (Eg primary user from sccm)

I appreciate the offer of testing and will definitely take you up on it.

**Comment by PrzemyslawKlys** [comment]
Any eta?

**Comment by ryannewington** [comment]
~2 weeks. UI is not my strong suit so it's taking be a bit longer than I would like.

________________________________

**Comment by sylwester-majcher** [comment]
Hi Guys,

Good news. I also wait!

Pozdrawiam
Sylwester Majcher
________________________________
Od: Ryan Newington <notifications@github.com>
Wysłane: Thursday, July 9, 2020 12:11:35 PM
Do: lithnet/laps-web <laps-web@noreply.github.com>
DW: Subscribed <subscribed@noreply.github.com>
Temat: Re: [lithnet/laps-web] Working in multiple domains in the same forest (#10)


~2 weeks. UI is not my strong suit so it's taking be a bit longer than I would like.

________________________________

**Comment by red-erik** [comment]
Hello Ryan,
we can't wait to test the new release (pre-compiled) :-)

Red.

**Comment by ryannewington** [comment]
You've all been very patient. It's getting very close to being done. I have been working on this non-stop to try and get it out the door. The beta version should be available very soon. Until then, here is a sneak peak of the new config UI

![image](https://user-images.githubusercontent.com/17229877/88652155-c817bb80-d10d-11ea-9277-22af9fbc629b.png)

**Comment by PrzemyslawKlys** [comment]
Take your time :-) looks great!

**Comment by red-erik** [comment]
Looks great, really.
Thank you in advance.

Red.

**Comment by sylwester-majcher** [comment]
Fantastic!

Pozdrawiam
Sylwester Majcher
________________________________
Od: red-erik <notifications@github.com>
Wysłane: Tuesday, July 28, 2020 12:23:36 PM
Do: lithnet/laps-web <laps-web@noreply.github.com>
DW: sylwester-majcher <sylwester.majcher@outlook.com>; Comment <comment@noreply.github.com>
Temat: Re: [lithnet/laps-web] Working in multiple domains in the same forest (#10)


Looks great, really.
Thank you in advance.

Red.

**Comment by ryannewington** [comment]
Hi all

We're getting very close to being ready for a beta. Thanks again for the patience here. I have a bit of a progress update,  as well as an ask of you.

First the progress update and remaining work:
- I'm currently working on the new installation process. There's a nice MSI based installation process now. 
- Then I'm turning NTLM off in my forests and testing to make sure everything works nicely with Kerberos, and there aren't any sneaky fallbacks to NTLM
- I've got a few changes to make to the web user interface. A bit of a style update and some new features.
- Finally, I have to get the getting started guides and documentation written up (at least the start of it so you can find your way around the app)

Once that is done, we'll be ready to open up for testing.

*Can I ask you to review the following page about how access control works in the new version, and tell me if you see any problems with the scenarios provided.*

https://github.com/lithnet/access-manager/wiki/Access-evaluation-in-Access-Manager-Service-(AMS)

You'll notice this is in a new github repository. I'm going to fork this one, as the new app is basically a complete re-write and has a bunch of new functionality. New name, new logo, etc. This repository will stay in place for the existing "LAPS Web App", until I deprecate it. The new version will be rebranded as "Lithnet Access Manager" and will live here -> https://github.com/lithnet/access-manager/

**Comment by PrzemyslawKlys** [comment]
LGTM ;)

**Comment by red-erik** [comment]
https://github.com/lithnet/access-manager/wiki/Access-evaluation-in-Access-Manager-Service-(AMS)
Clear and well described, nothing to add :-)
Waiting for the new beta to test it.
Thank you very much for all your hard work.

Red.

**Comment by ryannewington** [comment]
Thanks for the feedback everyone.

Here is a status update

- New web UI is complete
- Installers are complete
- Documentation is well under way, but still a bit of work to do here. You're welcome to browse [the documentation](https://github.com/lithnet/access-manager/wiki) that has been written so far to learn about key scenarios this product supports 
- I'm working on the 'first run' process at the moment, to guide the user through the set up process

It is at the stage where we've started our internal deployment to our environments for testing and validation of key scenarios ('dog-fooding' stage). A private beta if you will. Once that is done, public beta will open up.

Thanks again for your patience.

**Comment by red-erik** [comment]
https://github.com/lithnet/access-manager/wiki
Again, clear and well done. A suggestion could be to spent some more words on "Lithnet Access Manager Agent" vs LAPS Client.
Great work ;-)

Red.

**Comment by ryannewington** [comment]
Funny you mention that. I just added this page...

https://github.com/lithnet/access-manager/wiki/Choosing-a-local-admin-password-strategy

**Comment by red-erik** [comment]
Really well done !!! :-)

**Comment by ryannewington** [comment]
Hi all. We found an issue in testing late last week that has since been fixed, but we need to redo some of our tests early this week to make sure the issue is resolved. Everything else is ready to go, so we should have the beta out by the end of the week, provided no major issues are found. 🎉

**Comment by ryannewington** [comment]
Today is the day! The first public beta is out. Thankyou all for your patience in getting to this point. I know it has been a long wait, but hopefully you'll find the wait is worth it.

I'm going to wait a bit before publicly announcing the beta, just so I have bandwidth to help you guys through your implementation and any issues you may have first, since you've been waiting the longest. 

I'm really keen for feedback from those who can test. There is a lot in this version, but the LAPS support should be the most mature, given it's origins in laps-web, and its the place I'd expect everyone to start first. Please log any issues you find, be they unexpected behavior, unclear documentation, or suggestions for improvements. 

This is a big departure from the way things worked in LAPS web, so I suggest starting with having a read of the [documentation](https://github.com/lithnet/access-manager/wiki) and getting started guides first.

I'm going to close off this issue now, and ask that you [log any issues](https://github.com/lithnet/access-manager/issues) in the new access manager repository where we will work on them together there. 

https://github.com/lithnet/access-manager/releases/tag/v1.0.7596-beta.1
